mr/src/main/java/org/apache/iceberg/mr/hive/serde/objectinspector/IcebergDateObjectInspector.java

```diff
@@ -42,17 +42,17 @@ private IcebergDateObjectInspector() {
 
   @Override
   public Date getPrimitiveJavaObject(Object o) {
-    return o == null ? null : Date.valueOf((LocalDate) o);
+    return o == null ? null : Date.valueOf(o.toString());
   }
 
   @Override
-  public DateWritable getPrimitiveWritableObject(Object o) {
-    return o == null ? null : new DateWritable(DateTimeUtil.daysFromDate((LocalDate) o));
+  public DateWritableV2 getPrimitiveWritableObject(Object o) {
+    return o == null ? null : new DateWritableV2(DateTimeUtil.daysFromDate((LocalDate) o));
   }
 
   @Override
   public Object copyObject(Object o) {
-    return o == null ? null : new Date(((Date) o).getTime());
+    return o == null ? null : Date.valueOf(o.toString());
```

Comment:
Are you sure that `toString` is the appropriate method to call here to get the time from the date?

`toString` is usually unsafe in these situations, especially for a function that takes in `Object`. At the least, if an object were passed in previously that could not be casted to `java.sql.Date`, then the user would get a decent stack trace explaining that to them. 

To me, it seems like the stack trace's value to developers has decreased here, but perhaps I'm not understanding the semantics of `org.apache.hadoop.hive.common.type.Date.valueOf`.

Additionally, I think we tend to prefer to use `java.sql.Date` as it is the more portable representation across the many query engines that need to be supported. Additionally, I believe that `java.sql.Date` is expressed in the documentation as the physical type for a logical Date. Again, somebody please correct me if I'm wrong.

Reply:
Good catch!
The data object used by Hive internally for storing Date is changed, so instanceOf should be used to handle the differences between the two Date implementations.